File: src/rules/detect_unscoped_change.zig
const std = @import("std");
const rule = @import("rule.zig");
const symbol_extractor = @import("../util/symbol_extractor.zig");
const call_graph = @import("../util/call_graph.zig");

const NON_CODE_EXTENSIONS = [_][]const u8{
    ".md", ".rst", ".txt", ".json", ".yaml", ".yml", ".toml",
    ".lock", ".env", ".cfg", ".ini", ".css", ".scss",
    ".svg", ".png", ".jpg", ".gif", ".ico", ".woff", ".woff2",
};

const NON_CODE_PREFIXES = [_][]const u8{
    ".github/", "docs/",
};

const CodeFileEntry = struct { index: u16, file: rule.PrFile };

pub fn run(alloc: std.mem.Allocator, ctx: rule.RuleContext) ![]rule.RuleResult {
    const pr = switch (ctx.payload) {
        .pr => |p| p,
        .release => return passResult(alloc),
    };

    // Filter to code files with patches
    var code_files: std.ArrayList(CodeFileEntry) = .empty;
    for (pr.pr_files, 0..) |f, i| {
        if (f.patch == null) continue;
        if (isNonCodeFile(f.filename)) continue;
        try code_files.append(alloc, .{ .index = @intCast(i), .file = f });
    }

    // 0-1 code files: always scoped
    if (code_files.items.len <= 1) {
        return passResult(alloc);
    }

    // Extract symbols from each file
    var all_symbols: std.ArrayList(symbol_extractor.FileSymbols) = .empty;
    for (code_files.items) |entry| {
        const symbols = try symbol_extractor.extractSymbols(alloc, entry.file.filename, entry.file.patch.?);
        try all_symbols.append(alloc, symbols);
    }

    // Build call graph
    var graph: call_graph.CallGraph = .empty;

    // Create file-level nodes
    var file_nodes: std.ArrayList(call_graph.NodeId) = .empty;
    for (code_files.items, 0..) |entry, idx| {
        const node = try graph.addNode(alloc, entry.index, entry.file.filename, .file);
        try file_nodes.append(alloc, node);

        // Create definition nodes and merge with file node
        const symbols = all_symbols.items[idx];
        for (symbols.definitions) |def_name| {
            const def_node = try graph.addNode(alloc, entry.index, def_name, .function);
            graph.merge(node, def_node);
        }
    }

    // Cross-file edges: call-to-definition matching
    for (all_symbols.items, 0..) |syms_a, idx_a| {
        for (syms_a.calls) |call_name| {
            for (all_symbols.items, 0..) |syms_b, idx_b| {
                if (idx_a == idx_b) continue;
                for (syms_b.definitions) |def_name| {
                    if (std.mem.eql(u8, call_name, def_name)) {
                        graph.merge(file_nodes.items[idx_a], file_nodes.items[idx_b]);
                    }
                }
            }
        }
    }

    // Import edges: resolve import paths to changed files
    for (all_symbols.items, 0..) |syms_a, idx_a| {
        for (syms_a.imports) |import_path| {
            const resolved = resolveImport(import_path, code_files.items);
            if (resolved) |idx_b| {
                if (idx_a != idx_b) {
                    graph.merge(file_nodes.items[idx_a], file_nodes.items[idx_b]);
                }
            }
        }
    }

    // Count connected components
    const components = graph.componentCount();

    if (components <= 1) {
        return passResult(alloc);
    }

    // Build result with component details
    const severity: rule.Severity = if (components >= 3) .@"error" else .warning;

    const comp_groups = try graph.getComponents(alloc);

    var affected: std.ArrayList([]const u8) = .empty;
    var detail_buf: std.ArrayList(u8) = .empty;
    const writer = detail_buf.writer(alloc);

    for (comp_groups, 0..) |group, comp_idx| {
        try writer.print("  Component {d}:", .{comp_idx + 1});
        for (group) |file_idx| {
            const filename = pr.pr_files[file_idx].filename;
            try writer.print(" {s}", .{filename});
            try affected.append(alloc, filename);
        }
        try writer.print("\n", .{});
    }

    const message = try std.fmt.allocPrint(
        alloc,
        "PR has {d} disconnected change clusters",
        .{components},
    );

    const results = try alloc.alloc(rule.RuleResult, 1);
    results[0] = .{
        .rule_id = "detect-unscoped-change",
        .severity = severity,
        .message = message,
        .affected_files = try affected.toOwnedSlice(alloc),
        .suggestion = try detail_buf.toOwnedSlice(alloc),
    };
    return results;
}

fn passResult(alloc: std.mem.Allocator) ![]rule.RuleResult {
    const results = try alloc.alloc(rule.RuleResult, 1);
    results[0] = .{
        .rule_id = "detect-unscoped-change",
        .severity = .pass,
        .message = "PR is well-scoped",
        .affected_files = &[_][]const u8{},
        .suggestion = null,
    };
    return results;
}

fn isNonCodeFile(filename: []const u8) bool {
    for (NON_CODE_PREFIXES) |prefix| {
        if (std.mem.startsWith(u8, filename, prefix)) return true;
    }
    for (NON_CODE_EXTENSIONS) |ext| {
        if (std.mem.endsWith(u8, filename, ext)) return true;
    }
    return false;
}

fn resolveImport(import_path: []const u8, code_files: []const CodeFileEntry) ?usize {
    // Strip quotes (Go imports include them)
    var path = import_path;
    if (path.len >= 2 and (path[0] == '"' or path[0] == '\'')) {
        path = path[1 .. path.len - 1];
    }

    // Strip relative prefixes
    if (std.mem.startsWith(u8, path, "./")) {
        path = path[2..];
    } else if (std.mem.startsWith(u8, path, "../")) {
        // Can't resolve ../ without knowing the importing file's directory
        // Still try suffix matching
        path = path[3..];
    } else if (std.mem.startsWith(u8, path, "@/")) {
        path = path[2..];
    }

    // Convert Python dotted notation to path
    var path_buf: [512]u8 = undefined;
    var converted_path = path;
    if (std.mem.indexOf(u8, path, ".") != null and
        std.mem.indexOf(u8, path, "/") == null)
    {
        // Likely Python dotted import: foo.bar -> foo/bar
        var out_len: usize = 0;
        for (path) |ch| {
            if (out_len >= path_buf.len) break;
            path_buf[out_len] = if (ch == '.') '/' else ch;
            out_len += 1;
        }
        converted_path = path_buf[0..out_len];
    }

    // Match against changed file names (suffix match)
    var ext_buf: [512]u8 = undefined;
    // Copy converted_path into ext_buf so we can safely append extensions
    if (converted_path.len > ext_buf.len) return null;
    @memcpy(ext_buf[0..converted_path.len], converted_path);

    for (code_files, 0..) |entry, idx| {
        const fname = entry.file.filename;
        // Exact suffix match
        if (std.mem.endsWith(u8, fname, converted_path)) return idx;
        // Try with common extensions
        const extensions = [_][]const u8{ ".ts", ".tsx", ".js", ".jsx", ".py", ".go", "/index.ts", "/index.js" };
        for (extensions) |ext| {
            if (converted_path.len + ext.len > ext_buf.len) continue;
            @memcpy(ext_buf[converted_path.len .. converted_path.len + ext.len], ext);
            const with_ext = ext_buf[0 .. converted_path.len + ext.len];
            if (std.mem.endsWith(u8, fname, with_ext)) return idx;
        }
    }
    return null;
}

File: crates/cli/src/rules/detect_unscoped_change.rs
use anyhow::Result;
use gh_verify_core::scope::{
    FileRole, classify_file_role, classify_scope, extract_feature_namespace, is_non_code_file,
    resolve_import, should_bridge_aux_to_source, should_bridge_colocated_sources,
    should_bridge_fork_variants, should_bridge_patch_semantic_tokens,
    should_bridge_test_fixture_pair,
};
use gh_verify_core::union_find::{NodeKind, UnionFind};
use gh_verify_core::verdict::{RuleResult, Severity};

use crate::util::symbol_extractor;

use super::{Rule, RuleContext};

const RULE_ID: &str = "detect-unscoped-change";

pub struct DetectUnscopedChange;

impl Rule for DetectUnscopedChange {
    fn id(&self) -> &'static str {
        RULE_ID
    }

    fn run(&self, ctx: &RuleContext) -> Result<Vec<RuleResult>> {
        let pr_files = match ctx {
            RuleContext::Pr { pr_files, .. } => pr_files,
            RuleContext::Release { .. } => return Ok(vec![pass_result()]),
        };

        // Filter to code files with patches
        let code_files: Vec<(u16, &crate::github::types::PrFile)> = pr_files
            .iter()
            .enumerate()
            .filter(|(_, f)| f.patch.is_some() && !is_non_code_file(&f.filename))
            .map(|(i, f)| (i as u16, f))
            .collect();

        // 0-1 code files: always scoped
        if code_files.len() <= 1 {
            return Ok(vec![pass_result()]);
        }

        // Extract symbols from each file
        let mut all_symbols = Vec::new();
        for (_, file) in &code_files {
            let symbols =
                symbol_extractor::extract_symbols(&file.filename, file.patch.as_deref().unwrap())?;
            all_symbols.push(symbols);
        }

        let aux_count = code_files
            .iter()
            .filter(|(_, f)| classify_file_role(&f.filename) != FileRole::Source)
            .count();
        let source_count = code_files.len().saturating_sub(aux_count);

        // Build call graph
        let mut graph = UnionFind::new();

        // Create file-level nodes
        let mut file_nodes = Vec::new();
        for (idx, (file_idx, file)) in code_files.iter().enumerate() {
            let node = graph.add_node(*file_idx, &file.filename, NodeKind::File);
            file_nodes.push(node);

            // Create definition nodes and merge with file node
            let symbols = &all_symbols[idx];
            for def_name in &symbols.definitions {
                let def_node = graph.add_node(*file_idx, def_name, NodeKind::Function);
                graph.merge(node, def_node);
            }
        }

        // Cross-file edges: call-to-definition matching
        for (idx_a, syms_a) in all_symbols.iter().enumerate() {
            for call_name in &syms_a.calls {
                for (idx_b, syms_b) in all_symbols.iter().enumerate() {
                    if idx_a == idx_b {
                        continue;
                    }
                    for def_name in &syms_b.definitions {
                        if call_name == def_name {
                            graph.merge(file_nodes[idx_a], file_nodes[idx_b]);
                        }
                    }
                }
            }
        }

        // Import edges: resolve import paths to changed files
        let filenames: Vec<&str> = code_files
            .iter()
            .map(|(_, f)| f.filename.as_str())
            .collect();
        for (idx_a, syms_a) in all_symbols.iter().enumerate() {
            for import_path in &syms_a.imports {
                if let Some(idx_b) = resolve_import(import_path, &filenames) {
                    if idx_a != idx_b {
                        graph.merge(file_nodes[idx_a], file_nodes[idx_b]);
                    }
                }
            }
        }

        // Feature namespace pre-pass: bridge files sharing a dominant feature token.
        let paths_for_ns: Vec<&str> = code_files
            .iter()
            .map(|(_, f)| f.filename.as_str())
            .collect();
        if aux_count > 0 {
            if let Some(ns) = extract_feature_namespace(&paths_for_ns) {
                if ns.member_indices.len() >= 2 {
                    let anchor = file_nodes[ns.member_indices[0]];
                    for &idx in &ns.member_indices[1..] {
                        graph.merge(anchor, file_nodes[idx]);
                    }
                }
            }
        }

        // Guarded weak edges for semantics (source+test+fixture, colocated feature files).
        for idx_a in 0..code_files.len() {
            for idx_b in (idx_a + 1)..code_files.len() {
                let path_a = &code_files[idx_a].1.filename;
                let path_b = &code_files[idx_b].1.filename;

                let should_merge = should_bridge_colocated_sources(path_a, path_b)
                    || should_bridge_aux_to_source(path_a, path_b, source_count, aux_count)
                    || should_bridge_aux_to_source(path_b, path_a, source_count, aux_count)
                    || should_bridge_fork_variants(path_a, path_b)
                    || should_bridge_test_fixture_pair(path_a, path_b)
                    || should_bridge_patch_semantic_tokens(
                        path_a,
                        path_b,
                        &all_symbols[idx_a].identifiers,
                        &all_symbols[idx_b].identifiers,
                        source_count,
                        aux_count,
                    );

                if should_merge {
                    graph.merge(file_nodes[idx_a], file_nodes[idx_b]);
                }
            }
        }

        // Count connected components
        let components = graph.component_count();
        let severity = classify_scope(code_files.len(), components);

        if severity == Severity::Pass {
            return Ok(vec![pass_result()]);
        }

        // Build result with component details
        let comp_groups = graph.get_components();
        let mut affected = Vec::new();
        let mut detail = String::new();

        for (comp_idx, group) in comp_groups.iter().enumerate() {
            detail.push_str(&format!("  Component {}:", comp_idx + 1));
            for &file_idx in group {
                let filename = &pr_files[file_idx as usize].filename;
                detail.push_str(&format!(" {filename}"));
                affected.push(filename.clone());
            }
            detail.push('\n');
        }

        Ok(vec![RuleResult {
            rule_id: RULE_ID.to_string(),
            severity,
            message: format!("PR has {components} disconnected change clusters"),
            affected_files: affected,
            suggestion: Some(detail),
        }])
    }
}

fn pass_result() -> RuleResult {
    RuleResult::pass(RULE_ID, "PR is well-scoped")
}
